web/packages/teleterm/src/services/config/migration.ts

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright 2023 Gravitational, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { FileStorage } from 'teleterm/services/fileStorage';
+import Logger from 'teleterm/logger';
+
+const logger = new Logger('ConfigMigration');
+
+// TODO(gzdunek): when a migration fails an error should be thrown.
+// Currently, `FileStorage.write` catches all IO errors,
+// but instead they should be propagated and handled at higher levels.
+// If an error occurs, we should show it to the user and stop the app from working.
```

Comment:
Why would we have to stop the app from working? Couldn't we warn the user instead and continue the startup of the app?

I don't think we need to prescribe how the error should be handled. I'd focus on the fact that we need two modes of interaction with the file storage API, one that ignores the errors and one that doesn't. Here we'd definitely be interested in the errors so that we can react to that appropriately.

---

A note for the future: one idea that came to my mind was to investigate whether

* any usage of file storage on the Electron side needs to swallow errors
* any usage of file storage on the frontend side needs to inspect errors

If there's no need for any of those things, then perhaps this could be handled when doing IPC or sth.

Reply:
>Why would we have to stop the app from working? Couldn't we warn the user instead and continue the startup of the app?

Hmm okay, it may be too restrictive (especially in this case). I thought that working on a not-migrated state could cause problems, but OTOH the migration happens before the schema validation, so it is not possible to have the app config in a shape that is different from we expect.